src/backend/distributed/planner/local_distributed_join_planner.c

```diff
@@ -0,0 +1,566 @@
+ * and result relation.
+ */
+static RTEToSubqueryConverterReference *
+GetNextRTEToConvertToSubquery(FromExpr *joinTree,
+							  RTEToSubqueryConverterContext *rteToSubqueryConverterContext,
+							  PlannerRestrictionContext *plannerRestrictionContext, Oid
+							  resultRelationId)
+{
+	RTEToSubqueryConverterReference *localRTECandidate = NULL;
+	RTEToSubqueryConverterReference *distributedRTECandidate = NULL;
+	if (!FillLocalAndDistributedRTECandidates(rteToSubqueryConverterContext,
+											  &localRTECandidate,
+											  &distributedRTECandidate))
+	{
+		return NULL;
+	}
+
+	if (OidIsValid(resultRelationId))
+	{
+		if (resultRelationId == GetRTEToSubqueryConverterReferenceRelId(
+				localRTECandidate))
+		{
+			return distributedRTECandidate;
+		}
+		if (resultRelationId == GetRTEToSubqueryConverterReferenceRelId(
+				distributedRTECandidate))
+		{
+			return localRTECandidate;
+		}
+	}
+
+	if (LocalTableJoinPolicy == LOCAL_JOIN_POLICY_PREFER_LOCAL)
+	{
+		return localRTECandidate;
+	}
+	else if (LocalTableJoinPolicy == LOCAL_JOIN_POLICY_PREFER_DISTRIBUTED)
+	{
+		if (distributedRTECandidate)
+		{
+			return distributedRTECandidate;
+		}
+		else
+		{
+			return localRTECandidate;
+		}
+	}
+	else
+	{
+		if (ShouldConvertDistributedTable(joinTree, distributedRTECandidate))
+		{
+			return distributedRTECandidate;
+		}
+		else
+		{
+			return localRTECandidate;
+		}
+	}
+}
+
+
+/*
+ * FillLocalAndDistributedRTECandidates fills the local and distributed RTE candidates.
+ * It returns true if we should continue converting tables to subqueries.
+ */
+static bool
+FillLocalAndDistributedRTECandidates(
+	RTEToSubqueryConverterContext *rteToSubqueryConverterContext,
+	RTEToSubqueryConverterReference **localRTECandidate,
+	RTEToSubqueryConverterReference **
+	distributedRTECandidate)
+{
+	if (list_length(rteToSubqueryConverterContext->localTableList) > 0)
+	{
+		*localRTECandidate = linitial(rteToSubqueryConverterContext->localTableList);
+	}
+	if (*localRTECandidate == NULL)
+	{
+		return false;
+	}
+
+	if (list_length(rteToSubqueryConverterContext->distributedTableList) > 0)
+	{
+		*distributedRTECandidate = linitial(
+			rteToSubqueryConverterContext->distributedTableList);
+	}
+	return *distributedRTECandidate != NULL ||
+		   rteToSubqueryConverterContext->hasSubqueryRTE;
+}
+
+
+/*
+ * GetRTEToSubqueryConverterReferenceRelId returns the underlying relation id
+ * if it is a valid one.
+ */
+static Oid
+GetRTEToSubqueryConverterReferenceRelId(
+	RTEToSubqueryConverterReference *rteToSubqueryConverterReference)
+{
+	if (rteToSubqueryConverterReference &&
+		rteToSubqueryConverterReference->rangeTableEntry)
+	{
+		return rteToSubqueryConverterReference->rangeTableEntry->relid;
+	}
+	return InvalidOid;
+}
+
+
+/*
+ * RemoveFromRTEToSubqueryConverterContext removes an element from
+ * the relevant list based on the relation id.
+ */
+static void
+RemoveFromRTEToSubqueryConverterContext(
```

Comment:
this would be more meaningful / easier to understand if it was named something like `RemoveFromConversionCandidates`